ubuntu_image/classic_builder.py

```diff
@@ -0,0 +1,605 @@
+        # libparted's commit() operation resets type GUIDs to defaults and
+        # clobbers things like hybrid MBR partitions.
+        part_id = 1
+        for i, part in enumerate(volume.structures):
+            image.copy_blob(volume.part_images[i],
+                            bs=image.sector_size,
+                            seek=part.offset // image.sector_size,
+                            count=ceil(part.size / image.sector_size),
+                            conv='notrunc')
+            if part.role is StructureRole.mbr or part.type == 'bare':
+                continue
+            image.set_parition_type(part_id, part.type)
+            part_id += 1
+        for value, dest in offset_writes:
+            # Decipher non-numeric offset_write values.
+            if isinstance(dest, tuple):
+                dest = part_offsets[dest[0]] + dest[1]
+            image.write_value_at_offset(value // image.sector_size, dest)
+
+    def make_disk(self):
+        # Based on the -o/--output and -O/--output-dir options, and the volumes
+        # in the gadget.yaml file, we can now calculate where the generated
+        # disk images should go.  We'll write them directly to the final
+        # destination so they don't have to be moved later.  Here is the
+        # option precedence:
+        #
+        # * The location specified by -o/--output;
+        # * <output_dir>/<volume_name>.img
+        # * <work_dir>/disk.img
+        #
+        # If -o was given and there are multiple volumes, we ignore it and
+        # act as if -O is in use.
+        disk_img = None
+        if self.output is not None:
+            if len(self.gadget.volumes) > 1:
+                _logger.warn('-o/--output ignored for multiple volumes')
+            else:
+                disk_img = self.output
+        if not disk_img:
+            os.makedirs(self.output_dir, exist_ok=True)
+        # Walk through all partitions and write them to the disk image at the
+        # lowest permissible offset.  We should not have any overlapping
+        # partitions, the parser should have already rejected such as invalid.
+        #
+        # XXX: The parser should sort these partitions for us in disk order as
+        # part of checking for overlaps, so we should not need to sort them
+        # here.
+        for name, volume in self.gadget.volumes.items():
+            image_path = (
+                disk_img if disk_img is not None
+                else os.path.join(self.output_dir, '{}.img'.format(name)))
+            self._make_one_disk(image_path, name, volume)
+        self._next.append(self.generate_manifests)
+
+    def generate_manifests(self):
+        # After the images are built, we would also like to have some image
+        # manifests exported so that one can easily check what packages
+        # have been installed on the rootfs.
+        # We utilize dpkg-query tool to generate the manifest file.
+
+        deprecated_words = ['ubiquity', 'casper']
```

Comment:
Could you add a comment as to what/why those deprecated words exist?

I can see they are excluded from the manifest, but I don't know why.

Reply:
I refer to 
https://help.ubuntu.com/community/MakeALiveCD/DVD/BootableFlashFromHarddiskInstall
and remove relevant lines which contain those deprecated words from the manifest file.

Reply:
Ah, right, that makes sense.

Adding a comment with that URL and explanation would be great.